README.MD

<h1 align="center"> Personalized Health </h1>
<p align="center">
  <img src="objects/images/readme1.jpg" alt="Title-Image" />
</p>

<p align="center">
 <b>Robert Renecker - Jacob Munoz - Akshit Arora</b>
</p>

## Versions Selected

Swift 3 \
WatchOS 5.1.3 \
iOS 12.1.4 \
XCode 10

## The Idea and Mission Statement

The goal for our project is to build a personalized goal application for running. Given a broad spectrum ranging from never-have-run-before to I-run-every-day, a regression model will be able to provide inference to a recommendation that a potential runner should carry out in order to meet a goal. The goals may be weight loss or the ability to run a certain distance. By using this application, hopefully a potential runner will be able to achieve these goals in a way that is personalized to their health and ability.

The problem that we hope to solve is that right now, running for some people is difficult. However, most of the problem is just figuring out a realistic goal that will encourage them to attempt to run. By using this web-app, we hope to give them a realistic goal that may not be too difficult for them to do. Thus, it will encourage them to continue making progress towards a longer-term goal with which they feel proud of achieving.

As a team, we feel that this project is more than feasible given we understand our problem statement and we understand the data we are working with. The scope of our project isn’t too large and it encompasses minimal things which will make it possible to do with the time constraints we’ve been given.

In regards to our machine learning model, we believe that multiple linear regression will be a good interpreter for recommending goals to a potential runner because there is most likely a linear relationship between various features in ones health and that of their running ability.

The user will input some initial training data(how much are you currently running, age, weight, distance goal, more?) to provide some baseline for running experience. This will give them an initial running goal to attempt. The model will then collect data during the run(distance traveled vs. goal, heart rate, pace, more?) and adjust accordingly. We want our model to be able to adjust itself based on user performance. For example, if the user crushes the first goal given by the app, we want it to adjust and give the user a much more rigorous running plan. If the user struggled to reach the goal or didn’t complete the run, it should adjust accordingly to give a more reasonable goal. Ideally the app would underestimate rather than overestimate the user so as to not demotivate them!

changelog.md
### Week 9 (3/10 - 3/16) - Concept:
  - All:
    - Draft and finalize project proposal
    - Compound realistic goals with team
    - Talk with class and other groups; Feedback analysis
    - Storyboarding
    - Tried FitBit API, switched to iOS based on the feedback received and because FitBit only offered offline data

 ### Week 10 (3/17 - 3/23) - Planning:
  - [Personalized Health - Week 10 Video update](https://youtu.be/LEdRSpescg0)
  - Robert Renecker ([@robertrenecker](https://github.com/robertrenecker))
    - Download software needed
    - Plan out tools needed and find other similar projects done for reference.
    - Boiler Plate application written in Swift. Authorizes users on watchOS/iOS to let App use HealthKit. ([link to iRun](https://github.com/CUBoulder-2019Sp-IML4HCI/FinalProject-PersonalizedHealth/tree/master/iRun))
    - Research of Swift, Xcode, HealthKit API (tutorials + couple hours of reading material)
    - Youtube segment

  - Jacob Munoz ([@jamu0075](https://github.com/jamu0075))
    - Explored UI design tools (Sketch, Adobe XD, and Figma) and decided on Figma for it's ease of use without loss of capabilities
    - Discussed with Robert and Akshit the architecture of our project to better understand the deisgn capabilities of iOS development
    - Further researched iOS UI design through Apple's direct guidelines ([link](https://developer.apple.com/design/human-interface-guidelines/ios/overview/themes/))

  - Akshit Arora ([@aroraakshit](http://github.com/aroraakshit))
    - Completed Udacity course on iOS dev. Built a boilerplate app to understand iOS application lifecycle ([PitchPerfect branch - link](https://github.com/CUBoulder-2019Sp-IML4HCI/FinalProject-PersonalizedHealth/tree/pitchPerfect/pitchPerfect))
    - Studied architectures of sample projects on GitHub that used HealthKit to design an architecture for our app. ([Vimo HeartRate](https://github.com/CUBoulder-2019Sp-IML4HCI/FinalProject-PersonalizedHealth/tree/sampleWatchOS/watchOS-3-heartrate), [Diabetes Monitoring Assist](https://github.com/wytesk133/diabetes-monitoring-assist), [Aima](https://github.com/leonidprovorov/aima))
    - Finalized an architecture where the views for iOS app and WatchOS app are separate, and the logic to control the views (aka. model) for both the apps is separate
    - Finalized versions for XCode, iOS, WatchOS, Swift to avoid dev version conflicts

### Week 11 (3/24 - 3/30) - Spring Break

### Week 12 (3/31 - 4/6) - Design & Development:
- [Personalized Health - Week 12 Video update](https://youtu.be/pEmbwJ8PBj8)
- Robert Renecker ([@robertrenecker](https://github.com/robertrenecker))
  - Implemented the iOS application. Configured persistence of data that the user will submit on their phone. Helped take from the mock-ups to the design table in Xcode to connect these features to the design we are moving towards. Put together youtube video.

- Akshit Arora ([@aroraakshit](http://github.com/aroraakshit))
  - Explored CreateML and CoreML tutorials to understand workflows for ML in iOS development. To check whether online learning is possible or not. ([Better Rest branch - link](https://github.com/CUBoulder-2019Sp-IML4HCI/FinalProject-PersonalizedHealth/tree/createML), [Tutorial 1](https://www.youtube.com/watch?v=x9_BG2q7XYw), [Tutorial 2](https://www.youtube.com/watch?v=T4t73CXB7CU))
  - Drafted an architecture for ML models based on couch to 5k plan.

- Jacob Munoz ([@jamu0075](https://github.com/jamu0075))
  - Implemented more UX using figma to provide the key features of the application. The [mock-up](https://www.figma.com/proto/DUEEdmnHawuhMYVCu4EsruqB/Personalized-Health?node-id=1%3A2&scaling=scale-down) is intended to demo the applications interactions without a final GUI to ensure it is intuitive.

### Week 13 (4/7 - 4/13) - Design & Development:
- [Personalized Health - Week 13 Video update](https://youtu.be/XtAauaUkJPM)

- Akshit Arora ([@aroraakshit](http://github.com/aroraakshit))
  - Created first draft of our ML Model and coded a script using Apple's [CreateML](https://developer.apple.com/documentation/createml) to create a .mlmodel file.
  - Integrated the first iteration of ML model into iRun basic interface using a pre-trained model file (.mlmodel) through Apple's [CoreML](https://developer.apple.com/documentation/coreml).
  - Identified more features that can be used and explored options to have the model end-user trainable.

- Robert Renecker ([@robertrenecker](https://github.com/robertrenecker))
  - Iterated over UI. Adapted different Swift projects to use in alignment of our goal and UI design.
  - Incorporated machine learning model into iOS application that we will be using with the user.
  - Ran inference on model to test UI compatibility with feature output data.
  - Found / researched libraries for visualization methods that we want to incorporate into swift.

- Jacob Munoz ([@jamu0075](https://github.com/jamu0075))
  - Researched the Charts library for swift data visualizations to help Robert when we get there.
  - Worked with Robert and Akshit to define a MVP
  - Design a simple first iteration application page for accepting user running input and providing output
  - Began designing an application flow to 'gameify' the stages of running training to keep users motivated


### Week 14 (4/14-4/20) - Quality Checks:
- [Personalized Health - Week 14 Video update](https://youtu.be/2pR7CaKURUU)

- Robert Renecker ([@robertrenecker](https://github.com/robertrenecker))
  - Found available dataset to train model on
  - Cleaned dataset, synchronized with data available from user input in iOS application.
  - Data feature analysis.
  - Incorporated this model in MLKit. Trained given dataset found from Kaggles' Boston Marathon Runner Times 15'16'17 Dataset.
  - Updated UI to be more friendly to users
  - Updated UI to handle navigation from 1 main controller versus decoupled-navigation-controllers.
  - Youtube video
  - Live demo proof of concept


training_and_ML/dummy_data_script.py
import json
list_holder = []

#Create Objects
"""
    {'tot_dist': x, 'tot_time': y,'pace': something, 'recommendation': z}
"""
#Do increasing weight
distance = 1
for _ in range(100):
    for pace in range(5,60):
        rec = 0
        #Check the levels or recommendation levels
        if pace < 6:
            rec = 7
            pass
        elif pace < 7:
            rec = 7
            pass
        elif pace < 8:
            rec = 6
            pass
        elif pace < 10:
            rec = 5
            pass
        elif pace < 15:
            rec = 4
            pass
        elif pace < 17:
            rec = 3
            pass
        elif pace < 20:
            rec = 2
            pass
        elif pace < 25:
            rec = 1
            pass
        elif pace < 60:
            rec = 0
        list_holder.append({'pace':pace, 'distance':distance, 'rec':rec})
distance = 2
for _ in range(100):

    for pace in range(5,60):
        rec = 0
        #Check the levels or recommendation levels
        if pace < 7:
            rec = 8
            pass
        elif pace < 8:
            rec = 7
            pass
        elif pace < 9:
            rec = 7
            pass
        elif pace < 10:
            rec = 7
            pass
        elif pace < 15:
            rec = 5
            pass
        elif pace < 20:
            rec = 3
            pass
        elif pace < 30:
            rec = 2
            pass
        elif pace < 45:
            rec = 1
            pass
        elif pace < 60:
            rec = 0
        list_holder.append({'pace':pace, 'distance':distance, 'rec':rec})
distance = 3
for _ in range(100):

    for pace in range(5,60):
        rec = 0
        #Check the levels or recommendation levels
        if pace < 9:
            rec = 8
            pass
        elif pace < 10:
            rec = 7
            pass
        elif pace < 12:
            rec = 6
            pass
        elif pace < 15:
            rec = 5
            pass
        elif pace < 20:
            rec = 4
            pass
        elif pace < 30:
            rec = 3
            pass
        elif pace < 40:
            rec = 2
            pass
        elif pace < 45:
            rec = 1
            pass
        elif pace < 60:
            rec = 0
        list_holder.append({'pace':pace, 'distance':distance, 'rec':rec})



with open('training_running_data.json', 'w') as outfile:
    json.dump(list_holder, outfile)
